File: src/0416-partition-equal-subset-sum/solution.cpp
class Solution {
public:
    bool canPartition(vector<int>& n) {
        int s = accumulate(n.begin(), n.end(), 0), t = s >> 1;
        if (s & 1) return false;
        vector<bool> dp(t+1, false);
        dp[0] = true;
        for(auto nn : n) {
            for(int i = t; i >= nn; i--) {
                dp[i] = dp[i] || dp[i - nn];
            }
        }
        return dp[t];
    }
};

File: src/0416-partition-equal-subset-sum/solution.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        if sum(nums) % 2: return False

        dp, target = set(), sum(nums) // 2
        dp.add(0)

        for i in range(len(nums)):
            newDp = set(dp)
            for t in dp:
                if t + nums[i] == target: return True
                newDp.add(t + nums[i])
            dp = newDp
        return False

